Realm.Shared/linq/RealmResultsVisitor.cs

```diff
@@ -179,44 +203,70 @@ internal override Expression VisitMethodCall(MethodCallExpression m)
                             firstRowPtr = rh.GetRow(0);
                         }
                     }
-                    if (firstRowPtr == IntPtr.Zero)
+                    if (firstRowPtr != IntPtr.Zero)
+                        return Expression.Constant(_realm.MakeObjectForRow(_metadata, firstRowPtr));
+                    if (m.Method.Name == "First")
                         throw new InvalidOperationException("Sequence contains no matching element");
-                    return Expression.Constant(_realm.MakeObjectForRow(_metadata, firstRowPtr));
+                    Debug.Assert (m.Method.Name == "FirstorDefault");
+                    return Expression.Constant(null);
                 }
-                if (m.Method.Name == "Single")  // same as unsorted First with extra checks
+                if (m.Method.Name == "DefaultIfEmpty")
```

Comment:
I don't think this is how `DefaultIfEmpty` is supposed to behave - http://stackoverflow.com/questions/8854181/defaultifempty-in-linq